src/App/mermaidData/pullApiData.js

```diff
@@ -89,55 +79,40 @@ export const pullApiData = async ({
-          // those responses will include an array of ids to wipe out everything associated with the project
-          // causing the 401 or 403
-          const isAnyErrorCodeAssociatedRecordsToBeDeleted =
-            error && [401, 403].includes(error.code)
-          const errorCodeAssociatedIds = isAnyErrorCodeAssociatedRecordsToBeDeleted
-            ? error.record_ids
-            : []
+          const is401Or403 = error && [401, 403].includes(error.code)
+
           const isDataTypeProjectAssociated = getIsDataTypeProjectAssociated(apiDataType)
 
-          const bulkDeleteIdsWithNoDuplicates = Array.from(
-            new Set([...deleteIds, ...errorCodeAssociatedIds]),
-          )
+          const bulkDeleteIdsWithNoDuplicates = Array.from(new Set([...deleteIds, ...removeIds]))
 
           dexiePerUserDataInstance[apiDataType].bulkPut(updatesWithPushToApiTagReset)
           dexiePerUserDataInstance[apiDataType].bulkDelete(bulkDeleteIdsWithNoDuplicates)
 
-          if (isAnyErrorCodeAssociatedRecordsToBeDeleted && isDataTypeProjectAssociated) {
-            resetLastRevisionNumberForProjectDataType({
-              dataType: apiDataType,
-              projectId,
-              dexiePerUserDataInstance,
-            })
-          }
-        }
-
-        // If the user has been removed from a project, the backend does not treat
-        // it as a change to the project itself, therefore no revision is triggered.
-        // Only the project profile changes in this case.
-        // If the user remains online, project profile will not be updated by a pull.
-        // Determine if the user's project membership has changed based on the
-        // /projects endpoint response and delete the project in the IndexedDb if
-        // they have been removed.
-        if (projectsResponse && apiDataType === 'projects') {
-          const projectsResults = projectsResponse.data?.results
-
-          // Determine which projects in IndexedDB are not in the /projects API response
-          const deleteProjectIds = indexedDbProjects
-            .filter(
-              (indexedDBProject) =>
-                !isIndexedDBProjectInProjectResults(indexedDBProject, projectsResults),
-            )
-            .map((removedProject) => removedProject.id)
-
-          if (deleteProjectIds.length) {
-            // Delete the projects from IndexedDB as the user has been removed from them
-            dexiePerUserDataInstance.projects.bulkDelete(deleteProjectIds)
+          if (is401Or403 && isDataTypeProjectAssociated && projectId) {
+            // we still delete project related data in addition to anything in the removes array,
+            // because the removes array response doesnt include all the things we want to delete currently.
+            const deleteProjectRelatedRecords = dexiePerUserDataInstance[apiDataType]
+              .where({ project: projectId })
+              .delete()
+            const resetProjectRelatedLastRevisionNumbers =
```

Comment:
if a project is deleted/removed, are we resetting the last revision to one prior to deletion? is there a new revision number added upon deletion?

Reply:
this is the weird part. The `deletes` array is for data that is deleted as data. HEre, a user is just not allowed on the project. So nothing is `deleted` as data, but de do want to remove access from the user (so we do delete it from the browser). Next time that user pulls and they do have permission, they will be pulling everything from a fresh slate (because we deleted everything for them when they were denied), so we just null out their last revision number to tell the server to give us everything (everything relavent) for that project+user

Reply:
TBH, its a bit complicated for the front end. Ideally, the API would reset last revision numbers when a user is denied and just give us waht we need to remove in the `removes` array. The front end should only concern itself with storing whatever last revision number the server sends. But this is how it evolved/currently works

Reply:
ahh got it